node/cancellation-v2.md

```diff
@@ -0,0 +1,308 @@
+
+```ts
+async function main() {
+  await Context.cancelled;
+}
+
+let counter = 0;
+
+const signals = {
+  increment() {
+    ++counter;
+  },
+};
+// ...
+```
+
+### New concepts
+
+#### Recap from the documentation site (updated for this proposal)
+
+Temporal Workflows have different types that can be cancelled:
+
+- The entire Workflow
+- A timer or an Activity
+- Cancellable contexts created with `Context.cancellable`
+- [`Trigger`](#trigger-or-resolvablepromise)s
+
+Workflows are represented internally by a tree of contexts where the `main` function runs in the root context.
+Cancellation propagates from outer contexts to inner ones and is handled by catching `CancellationError`s when `await`ing on promises.
+
+#### `ContextualPromise`
+
+- Implements the `PromiseLike` interface (awaitable via `then()`)
+- Context aware: linked to parent context and propagates cancellation, transferable between different contexts
+- Gets associated with a context once awaited
+- Returned by `Context.nonCancellable`
+
+#### `CancellablePromise`
+
+- Extends `ContextualPromise`
+- Exposes the `cancel` method which when called will cancel Activities, timers and `Triggers` (see below)
+  - When cancelled these will throw a `CancellationError` unless they've already complete
+- Returned by Activities, timers and `Context.cancellable`
+
+#### `Trigger` (or `ResolvablePromise`)
+
+- Extends `CancellablePromise`
+- Provides additional public methods to resolve and reject itself
+- Useful for e.g waiting for a signal from WF main
+
+  ```ts
+  const unblocked = new Trigger<void>();
+  const signals = {
+    unblock(): void {
+      unblocked.resolve();
+    },
+  };
+
+  async function main() {
+    await unblocked;
+  }
+  ```
+
+### Examples
+
+#### Handle Workflow cancellation by an external client while an Activity is running
+
+```ts
+import { CancellationError } from '@temporalio/workflow';
+import { httpGetJSON } from '@activities';
+
+export async function main(url: string) {
+  let result: any = undefined;
+  try {
+    result = await httpGetJSON(url);
+  } catch (e) {
+    if (e instanceof CancellationError) {
+      console.log('Workflow cancelled');
+      // Cleanup logic goes here
+    } else {
+      throw e;
+    }
+  }
+  return result; // Will be undefined in case the Workflow was cancelled
+}
+```
+
+#### Cancel a timer from Workflow code
+
+```ts
+import { CancellationError, sleep } from '@temporalio/workflow';
+
+export async function main() {
+  // Timers and Activities are automatically cancelled when their promise is cancelled.
+  // Awaiting on a cancelled promise with throw the original CancellationError.
+  const promise = sleep(1);
+  promise.cancel(); // promise is a CancellablePromise
+  try {
+    await promise;
+  } catch (e) {
+    if (e instanceof CancellationError) {
+      console.log('Exception was propagated 👍');
+    } else {
+      throw e;
+    }
+  }
+}
```

Comment:
what is the developer experience if i forget to use try/catch?


```
export async function main() {
  const promise = sleep(1);
  promise.cancel();
  await promise; // what happens here
}
```

Reply:
The promise will throw a `CancellationError` causing the workflow to fail.